runtime/interface.cpp

```diff
@@ -506,68 +509,89 @@ shutdown_function_type *shutdown_functions = reinterpret_cast<shutdown_function_
 shutdown_functions_status shutdown_functions_status_value = shutdown_functions_status::not_executed;
 jmp_buf timeout_exit;
 bool finished = false;
-bool flushed = false;
 bool wait_all_forks_on_finish = false;
 
 } // namespace
 
-void f$fastcgi_finish_request(int64_t exit_code) {
-  if (flushed) {
-    return;
+const string_buffer * compress_http_query_body(string_buffer * http_query_body) {
+  php_assert(http_query_body != nullptr);
+
+  if (is_head_query) {
+    http_query_body->clean();
+    return http_query_body;
+  } else {
+    if ((http_need_gzip & 5) == 5) {
+      header("Content-Encoding: gzip", 22, true);
+      return zlib_encode(http_query_body->c_str(), http_query_body->size(), 6, ZLIB_ENCODE);
+    } else if ((http_need_gzip & 6) == 6) {
+      header("Content-Encoding: deflate", 25, true);
+      return zlib_encode(http_query_body->c_str(), http_query_body->size(), 6, ZLIB_COMPRESS);
+    } else {
+      return http_query_body;
+    }
   }
+}
 
-  if (!run_once) {
-    exit_code = 0; // TODO: is it correct?
+void ob_flush_into_system() {
+  php_assert (ob_cur_buffer >= 0);
+  for (int i = ob_system_level + 1; i <= ob_cur_buffer; i++) {
+    oub[ob_system_level].append(oub[i].c_str(), oub[i].size());
   }
+}
 
-  flushed = true;
+void f$flush() {
+  php_assert(ob_cur_buffer >= 0);
+  php_assert(active_worker != nullptr && active_worker->conn != nullptr);
 
-  php_assert (ob_cur_buffer >= 0);
-  int first_not_empty_buffer = 0;
-  while (first_not_empty_buffer < ob_cur_buffer && oub[first_not_empty_buffer].size() == 0) {
-    first_not_empty_buffer++;
+  string_buffer const * http_body = compress_http_query_body(&oub[ob_system_level]);
+  string_buffer const * http_headers = nullptr;
+  if (!flushed_http_connection) {
+    http_headers = get_headers();
+    flushed_http_connection = true;
   }
+  http_send_immediate_response(http_headers ? http_headers->buffer() : nullptr, http_headers ? http_headers->size() : 0,
+                               http_body->buffer(), http_body->size());
+  oub[ob_system_level].clean();
+  static_SB_spare.clean();
+}
 
-  for (int i = first_not_empty_buffer + 1; i <= ob_cur_buffer; i++) {
-    oub[first_not_empty_buffer].append(oub[i].c_str(), oub[i].size());
```

Comment:
Now it flushed to zero indexed buffer, before that it was `first_not_emty_buffer`. Why is it changed? Won't it break something?

Reply:
Good observation, fixed it